---
title: "R Notebook"
output: html_notebook
---


To keep consistent with the eQTL estimates (because GWAS and eQTL sometimes report different reference and alternate alleles), I switch the reference and effect alleles so that the effect allele occurs before the reference allele alphabetically. (this is weird, but it was easy to implement)


First let's choose which genes we want based on how well they performed in PrediXcan cross-validation (some good, some bad). 
```{r}
library(dplyr)
library(RColumbo)
library(feather)
gtexdbf <- "~/Downloads/PredictDB_Covariance/TW_Whole_Blood/TW_Whole_Blood_0.5.db"

gtexdb <- src_sqlite(gtexdbf,create = F)
gtex_genes <- tbl(src = gtexdb,"extra") %>% arrange(pred.perf.qval) %>% collect(n=Inf) %>% mutate(fgeneid=as.integer(gsub("ENSG([0-9]+).+","\\1",gene)))
```


#Selecting Genes
I've made 4 p-value significance regions from which I sample 1 gene each
```{r}
set.seed(6022)
gtex_sample_genes <- gtex_genes %>% mutate(cut_level=cut(pred.perf.pval,breaks = c(0,1e-5,0.05,0.1,0.9))) %>% 
  group_by(cut_level) %>% sample_n(1) %>% ungroup()
gtex_sample_genes
```

# Cis Results

**all** cis-eQTL results for those genes (stored in a big HDF5 file)
```{r}
gtex_marg_h5 <- "/home/nwknoblauch/Desktop/eQTL/Snake/Whole_Blood_Analysis.v6p.all_snpgene_pairs.h5"
gtex_cis_marg <- read_h5_df(gtex_marg_h5,groupname = "cis_eQTL",subcols = c("fgeneid","chrom","pos","weight","slope_se")) %>% 
  rename(thetahat=weight,se_thetahat=slope_se)
gtex_cis_marg <- inner_join(gtex_cis_marg,gtex_sample_genes,by="fgeneid") %>% select(chrom,se_thetahat,thetahat,gene,genename)
head(gtex_cis_marg)
write.table(gtex_cis_marg,"~/Dropbox/RColumbo/analyses/eqtl_gwas_sample/GTEx_WB_cis.tsv",col.names = T,row.names=F,sep="\t",quote=F)
```


trans results for those genes (stored in 22 HDF5 files)
```{r}
gtex_marg_trans_h5s <- paste0("/media/nwknoblauch/Data/GTEx/GTEx_v6p_h5files/ortho_flip/WB_Chr",1:22,"_v6p_ortho_flip.h5")
gtex_trans <- bind_rows(
  lapply(
    gtex_marg_trans_h5s,
    function(x,subset_df){
      return(read_h5_df(x,groupname = "trans_eQTL") %>% semi_join(subset_df,by="fgeneid"))
    },
    subset_df=gtex_sample_genes
    )
  )
gtex_trans <- rename(gtex_trans,thetahat=theta,se_thetahat=serr)
gtex_trans <- select(gtex_sample_genes,gene,genename,fgeneid) %>% inner_join(gtex_trans,by="fgeneid") %>% select(-fgeneid)

write.table(gtex_trans,"~/Dropbox/RColumbo/analyses/eqtl_gwas_sample/GTEx_WB_trans_truncated.tsv",col.names=T,row.names=F,quote=F,sep="\t")
```


GWAS for a midsize chromosome
Chromosme 12 is nicely sized.


```{r}

gwasf <- "/media/nwknoblauch/Data/gwas_data/iibdgc-trans-ancestry-summary-stats/EUR.IBD.gwas.assoc_2.feather"
gwas_data <- read_feather(gwasf)
gwas_data <- rename(gwas_data,pvalg=p.value,se_betahat=serr.g)
gwas_data <- mutate(gwas_data,betahat=ifelse(ref_allele<eff_allele,log(1/OR),log(OR))) %>% select(-eff_allele,-ref_allele)
gwas_data <- filter(gwas_data,chrom==12)
write.table(gwas_data,"~/Dropbox/RColumbo//analyses/eqtl_gwas_sample/IBD_GWAS_Chr12.tsv",sep="\t",col.names=T,row.names=F,quote=F)
```





